Doc/regions.md

```diff
@@ -0,0 +1,567 @@
+```C
+static int
+frame_setlineno(PyFrameObject *f, PyObject* p_new_lineno, void *Py_UNUSED(ignored))
+```
+This seems very bad to call from another thread/interpreter.
+
+Ultimately, a frame object should never escape the thread/interpreter that it was created in.
+
+
+
+## Regions in Veronapy.
+
+The previous section illustrates the many temporary objects that Python creates to handle variables.
+To support this, we need our region object model to support complex local structures for indexing into regions.
+To achieve this, we will use three kinds of object ownership:
+
+* Local object 
+* Immutable
+* Regions
+
+The "local objects" are used to represent the many local structures that Python creates such as function objects, frame objects and cells.
+These objects are typically not shared between threads, so normally should not require any region tracking.
+However, we need to support local objects becoming immutable to allow them to be shared, and suppport for groups of objects to be moved between interpreters.
+
+To pass mutable states between threads, we need to introduce a new construct to Python: "regions".
+A region is a group of objects that can move together across threads
+and can be accessed by at most one interpreter at a time.
+This removes the need for locks to protect the objects in the region.
+The region can be passed between threads, and the receiving interpreter can access the objects in the region.
+
+To efficiently support regions, we impose the following invariants that the runtime is responsible for maintaining:
+
+* Local objects can reach anything
+* Immutable objects can only reach other immutable objects.
+* Objects in a region can only reach other objects in the same region, immutable objects, or linear region entry points.
+* Regions track the number of incoming references from local objects.
+
+New objects are always created as a local object.
+If Veronapy features are not used, then all the objects in the system will be local objects and the runtime will behave as normal.
+
+
+### Regions
+
+Regions are used to pass mutable state between concurrent units of execution (behaviours).
+For usability, we need references into the region, that are not part of the region itself.
+These might be from the objects such as the frame, from closures or from function objects.
+This is precisely what local objects are for as they can refer into the region.
+
+The core challenge is allowing local objects to refer into a region,
+while allowing the region to be sent to another interpreter.
+To do this, we need to track the number of references from local objects to a region, the "local reference count" (LRC).
+
+When we send a region to another behaviour we need to ensure that the LRC is zero.
+To make this more usable, if the LRC is not zero, we can scan the current frame to find all references to the region, and remove them.
```

Comment:
"remove them" sounds scary. Suppose there's a region R1 containing objects o1 and o2 (amongst others). Now suppose there's a local variable x in some frame that points to o2. Are you saying that this operation just clears the frame's x? (If so, you probably have to set it to None rather than NULL, since there's some static analysis in the bytecode compiler that determines which variables are never NULL and generates different code for them.)

Reply:
This is one area where I am uncertain on what will lead to the nicest programming model, and I think we really need to build and try it.

Explicit setting to `None` (or even a `Dangle` type), would lead to things being easy to write, but could lead to unexpected behaviour. I think building the right debugging support would be essential here.

Perhaps, the Dangle type could carry a stack trace or code location, so that access can say, 
>  This reference was "Dangled" on line N of file.py.

Reply:
I just realized it's worse -- it can break deep assumptions in Python. If I have a local tuple that references something that's being moved into a region, you can't just patch the tuple -- tuples themselves are immutable (just not *deeply*). In such a case you might have to abort the move operation.

Reply:
745c0a5faf7391362d0f2393dea0f5f4d963e269